Get iris-grib working with latest versions of its dependencies (Iris, eccodes, cftime, cartopy, proj)

lbdreyer opened the pull request:
The v0.17.x branch has become quite stale. The lockfile that defines the CI environment had not been updated so was still using older versions of packages. I have since updated the lockfile and noticed a few test failures that this PR addresses.

I have split the commits so they each address a separate point:
* 9411bc9863f88dbae8db1e3b9aab143bc8422083 - Updates the lockfile (run this morning), moves the lockfiles dir locations from `requirements/nox.lock` to `requirements/ci/nox.lock`, and drops PY37 (which was dropped by Iris 3.2)
* 9af1b24705eb9d707655f2e83ed0905c2589e22a- Since cftime 1.4, `cftime.date2num` returns integers rather than floats so some cml needed updating:
```
>>> cftime.__version__
'1.3.1'
>>> date = datetime(2013, 2, 1)
>>> unit = 'hours since 1970-01-01 00:00:00'
>>> calendar = 'gregorian'
>>> cftime.date2num(date, unit, calendar)
377688.0

>>> cftime.__version__ 
'1.4.0'
>>> date = datetime(2013, 2, 1)
>>> unit = 'hours since 1970-01-01 00:00:00'
>>> calendar = 'gregorian'
>>> cftime.date2num(date, unit, calendar)
377688
```
* fc7a2217f6a1922bcab13e5483c74e22a9a483cd - Since eccodes 2.24, if you try to set an  unsigned key to be a negative value it raises an error. I have replaced whether we used `gribapi.set_*(grib_message, <KEY>, -1)` to instead set it to either GRIB_MISSING_LONG or 255 for bytes.
* e6699459e907597634526d895c954135596e6551- Since cartopy 0.20 and proj 8.0.1, `iris.coord_systems.Mercator().as_Cartopy_crs().transform_point(lon_0, lat_0, ccrs.Geodetic) returns _very slightly different values to I had to fix up the magic number we use in a test for Grid Definition Template 10 (which uses that)
* 052565763f7c10099a61f23e981ce5eef7fb18e4 - Since Iris 3.2, when loading in NAME files in the iris-grib tests, the cube that gets loaded in sometimes has an extra attribute, and sometimes creates a new z coordinate. I have adjusted the test to deal with this.

lbdreyer commented:
This is currently pointing at the v0.17.x branch, but we are going to do the mergeback of v0.17.x onto main first (#289). Once that is in, I will rebase onto main and instead point this at main instead.

I will put it into draft until then to ensure this doesn't accidentally get merged

pp-mo commented:
Hi @lbdreyer do you think this casts light on #277 -- whether we should just drop testing against latest-iris ?
If we have effectively concluded that testing against iris dev-version **_is_** worth the effort, then maybe #277 should be retired.

lbdreyer commented:
This PR now bases of the `main` branch. I updated the first commit with a new, refreshed lockfile.

lbdreyer commented:
> Hi @lbdreyer do you think this casts light on #277 -- whether we should just drop testing against latest-iris ?
> If we have effectively concluded that testing against iris dev-version **_is_** worth the effort, then maybe #277 should be retired.

I'm in two minds about this. Is there much value in catching breaking Iris changes early if it means we are constantly breaking iris-grib's CI. Or could we rather just manually test against `latest-iris` as the need arises. 
I'd like a little more time to think about this. For now though, I think we can keep testing against main

lbdreyer commented:
Coincidentally eccodes just released version 2.25 although it's not yet on conda-forge. That being said I'd still like to get this PR in then we can address any potential issues with 2.25 next week/when it is available

pp-mo commented on iris_grib/_save_rules.py:
```diff
@@ -15,6 +15,7 @@
 
 import cf_units
 import gribapi
+from gribapi import GRIB_MISSING_LONG
```
I note that is it possible to use `gribapi.grib_set_missing` for this.
Do you think that might be preferable @lbdreyer ?

N.B. I have now read up about this, and not all keys can take a 'missing' value.
Notably, some parameters defined by a code-table, typically a single-byte value, may need to be set to 255 for 'not present'.
This, IIUC is not a true 'missing' value but a table-defined value which happens to _mean_ that this param is unused.  And equally, some params (tables) have no such valid value.
This in in the ecmwf documentation FAQs as ["Why can't I set typeOfFirstFixedSurface to 'missing'?"](https://confluence.ecmwf.int/pages/viewpage.action?pageId=212440369)
and ["How can I find out which keys can be set to 'missing'?"](https://confluence.ecmwf.int/pages/viewpage.action?pageId=235610299)

pp-mo reviewed (changes requested):
Can't see anything wrong, but a couple of things where I thought it might be worth considering an alternative.  See what you think

lbdreyer commented:
I have updated this PR addressing some of the comments @pp-mo raised, particularly this comment:

Changes:
* Reran the lockfile generation to pick up eccodes 2.25 - no new test failures due to this
* Cherry-picked @pp-mo's commit casting units.date2num to float the: https://github.com/SciTools/iris-grib/pull/288/commits/519ca5e6ab797def9f5457e17c4ca0d578f73df8

I _haven't_ included changes to using grib_set_missing as I found issues with that as we discussed over the phone

lbdreyer commented on iris_grib/_save_rules.py:
```diff
@@ -15,6 +15,7 @@
 
 import cf_units
 import gribapi
+from gribapi import GRIB_MISSING_LONG
```
I'm not having much luck with gribapi.grib_set_missing. It's fine for the changes in `shape_of_the_Earth` but fails when used in `set_fixed_surfaces`. If I do:

```
        gribapi.grib_set_missing(grib, "typeOfSecondFixedSurface")
        gribapi.grib_set_missing(grib, "scaleFactorOfSecondFixedSurface")
        gribapi.grib_set_missing(grib, "scaledValueOfSecondFixedSurface")
```
I get errors that look like:

```
ERROR: test_stats_type_min (iris_grib.tests.unit.save_rules.test_product_definition_template_8.Test_type_of_statistical_processing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../lib/python3.8/unittest/mock.py", line 1325, in patched
    return func(*newargs, **newkeywargs)
  File ".../iris_grib/tests/unit/save_rules/test_product_definition_template_8.py", line 59, in test_stats_type_min
    product_definition_template_8(cube, grib)
  File ".../iris_grib/_save_rules.py", line 1246, in product_definition_template_8
    _product_definition_template_8_10_and_11(cube, grib)
  File ".../iris_grib/_save_rules.py", line 1291, in _product_definition_template_8_10_and_11
    product_definition_template_common(cube, grib, full3d_cube)
  File ".../iris_grib/_save_rules.py", line 1205, in product_definition_template_common
    set_fixed_surfaces(cube, grib, full3d_cube)
  File ".../iris_grib/_save_rules.py", line 927, in set_fixed_surfaces
    gribapi.grib_set_missing(grib, "typeOfSecondFixedSurface")
  File ".../lib/python3.8/site-packages/gribapi/gribapi.py", line 1672, in grib_set_missing
    h = get_handle(msgid)
  File ".../lib/python3.8/site-packages/gribapi/gribapi.py", line 165, in get_handle
    h = ffi.cast("grib_handle*", msgid)
  File ".../lib/python3.8/site-packages/cffi/api.py", line 300, in cast
    return self._backend.cast(cdecl, source)
TypeError: an integer is required
```

Even if I set the `typeOfSecondFixedSurface` to be 255, and use grib_set_missing for the other options:
```
        gribapi.grib_set(grib, "typeOfSecondFixedSurface", 255)
        gribapi.grib_set_missing(grib, "scaleFactorOfSecondFixedSurface")
        gribapi.grib_set_missing(grib, "scaledValueOfSecondFixedSurface")
```

I get errors:
```
======================================================================
ERROR: test_stats_type_min (__main__.Test_type_of_statistical_processing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../lib/python3.8/unittest/mock.py", line 1325, in patched
    return func(*newargs, **newkeywargs)
  File ".../iris_grib/tests/unit/save_rules/test_product_definition_template_8.py", line 59, in test_stats_type_min
    product_definition_template_8(cube, grib)
  File ".../iris_grib/_save_rules.py", line 1238, in product_definition_template_8
    _product_definition_template_8_10_and_11(cube, grib)
  File ".../iris_grib/_save_rules.py", line 1283, in _product_definition_template_8_10_and_11
    product_definition_template_common(cube, grib, full3d_cube)
  File ".../iris_grib/_save_rules.py", line 1197, in product_definition_template_common
    set_fixed_surfaces(cube, grib, full3d_cube)
  File "...iris_grib/_save_rules.py", line 928, in set_fixed_surfaces
    gribapi.grib_set_missing(grib, "scaleFactorOfSecondFixedSurface")
  File ".../lib/python3.8/site-packages/gribapi/gribapi.py", line 1672, in grib_set_missing
    h = get_handle(msgid)
  File ".../lib/python3.8/site-packages/gribapi/gribapi.py", line 165, in get_handle
    h = ffi.cast("grib_handle*", msgid)
  File ".../lib/python3.8/site-packages/cffi/api.py", line 300, in cast
    return self._backend.cast(cdecl, source)
TypeError: an integer is required

```

pp-mo replied:
Ok, well it seems that it is **_not_** possible to use grib_set_missing everywhere you might expect to.
  \- not even all of the 'scaleFactor' and 'ScaledValue' pairs relating to fixed surfaces
So I think if we _can't_ use this consistently, then it's simply not worth it. 😞

pp-mo commented:
Thanks @lbdreyer for persisting with this !
This should bring a new release quite close now.
